### ISSUE TITLE
Composition / Inheritence vs Inheritence thoughts

### ISSUE THREAD

**Issue by ProLoser** [ryanmcdermott/clean-code-javascript#135]
So I thought I'd personally share some philosophies that helped me personally distinguish when to use one over the other. Initially as a novice developer I loved convention over configuration and therefore loved inheritance, but nowadays I've come to greatly appreciate composition. However I've realized there's a very concrete reason for this:

**Frameworks benefit from convention.**
**Libraries benefit from composition.**

I found that _in a framework, it's acceptable to build the assumption / expectation that fellow developers will have to eventually learn_ all the conventions of your framework in order to leverage it and move quickly. It is almost a form of 'environment' in which you choose to develop in, and in so doing, the conventions allow you to move quickly and create reusable, shareable modules with little to no learning curve across modules. However, this is an expensive requirement as there is a greater learning curve for new engineers joining the project / framework for them to gain what is essentially "common knowledge" of the conventions in the framework. Simply looking at code in angular, ember, and react you see properties of objects that are not quite self-explanatory on their own or you may seek functionality or behavior that requires a certain degree of knowledge of the framework in order to solve. But once this knowledge is common you can rapidly develop and focus on larger scale problems.

_When developing a library, you can't expect knowledge of your conventions._ Libraries are far more likely to be moved and used and added and removed.  Composeable code creates a sort of self-documenting approach to someone joining the codebase or picking up the code. It becomes clearer to see how or why a library or object is behaving in a specific way because the behavior configuration is laid out before you vs being handled using hand-waving magic under the hood. All behavior is explicitly configured and created. This is handy for when a developer may only briefly interact with your code such as a library in order to achieve what they need and then will likely move on to focus on unrelated code (such as another section of their application).

### TL;DR:

A possible example of inheritence / convention vs composeable / configuration is are you developing a framework vs a library.

Is the user entrenching themself completely in your codebase and expected to learn all the conventions, or is their interaction with your code only used for a brief subset of their application.

Composeable is self-documenting and can be gleamed from the surface quickly. Inheritence requires gaining knowledge of internal behavior but allows _large_ productivity rapidly.

**Comment by ryanmcdermott** [comment]
Dean I think you have good way of explaining things. If you have the chance in the future there's a PR regarding settling the finer details of composition, inheritance, and classes in JS. I would love to get your help! https://github.com/ryanmcdermott/clean-code-javascript/pull/139

**Comment by ProLoser** [comment]
In retrospect after reading everything I wrote, I feel I may have to adjust my statement.

I don't think Inheritance and Composability are mutually exclusive. I have libraries I've developed that heavily leverage classes and inheritance internally (and externally if you develop plugins for it). However, to consume the library you use it in a composable way. Additionally, the code internally can be organized in a composable while still leveraging inheritance.
